### PR TITLE
SITL: Add drag simulation for Sub

### PR DESCRIPTION
**Pull request by patrickelectric**
This improves submarines simulation to calculate drag forces in linear and angular motion, creating a better simulation environment for test.

### CODE REVIEW

**Comment by tridge** [comment]
nice! Just need to get it passing CI tests, or adjust CI tests. It's failing some sub tests now

File: libraries/SITL/SIM_Submarine.cpp [hunk 1]
```diff
@@ -70,20 +70,39 @@ void Submarine::calculate_forces(const struct sitl_input &input, Vector3f &rot_a
...
+ * @param velocity Body frame velocity of fluid
+ * @param drag_coefficient Drag coefficient of body
+ * @param force Output forces
+ * $ F_D = rho * V^2 * A * C_D / 2 $
+ * rho = water density (kg/m^3), V = volume (m^3), A = area (m^2), C_D = drag_coefficient
+ */
+void Submarine::calculate_drag_force(const Vector3f &velocity, const Vector3f &drag_coefficient, Vector3f &force)
+{
+    const Vector3f velocity_2(
+        abs(velocity.x) * velocity.x,
```
Comment (jaxxzer) [inline libraries/SITL/SIM_Submarine.cpp]: please comment why you are maintaining the sign instead of squaring it properly as the equation above.

**Comment by patrickelectric** [comment]
So, it runs the test fine locally and not in the CI, what should we do about it ?
https://github.com/ArduPilot/ardupilot/issues/10910

**Comment by peterbarker** [comment]
So travis is not currently testing your head commit, which is really, really strange.

Travis seems to be telling me it is testing " Commit e3a0fc9" - and assuming that's not a post-rebase hash, that's not the same thing as your commit here.
 
I was able to reproduce the same error travis is showing on an older version of your PR.  I wonder if you rebase this and push it up again whether travis will then pass?  Your current version passes for me locally - even rebased.

Try a rebase and push it up?

**Comment by patrickelectric** [comment]
@peterbarker I just did the rebase and it appears that it's not working. 
I don't have idea what is happening here, I'll create a second PR just to see if this can fix the problem.

**Comment by patrickelectric** [comment]
Yes, both PR results in the same problem.

**Comment by peterbarker** [comment]
OK, this is a weird one.  Changing yaw means we no longer go forwards?  That's odd...

image

You can see we start yawing when the yaw input changes - but then don't stop...

Is it possible we have no rotational friction now?!

The map shows us going in circles, you see:
image

**Comment by peterbarker** [comment]
Gah.  To be clear, we keep going forward but also keep turning when the input is cancelled.

**Comment by peterbarker** [comment]
... RC outputs seem to indicate we're no longer thrusting, too.

**Comment by patrickelectric** [comment]
Is this happening only with this PR on in master branch ?
I did some tests here and in the first version I forgot to add the rotational drag,  that was corrected some time ago.
https://github.com/ArduPilot/ardupilot/pull/10701/files#diff-6c79088e2de21c5e3ab9572d0f7b4096R83

With my tests with sitl and manual control via QGC I was able to see the drag with angular and linear velocities while testing this PR and the same thing is happening in my local tests.

Are this outputs and plots from travis or from your local test ?
It's a bit late here and I'll not be able to test and take a look right now.
I also need to finish a paper until sunday, but I can take a look if you point how can I replicate this behaviors.

**Comment by peterbarker** [comment]
On Fri, 29 Mar 2019, Patrick José Pereira wrote:

> Is this happening only with this PR on in master branch ?

I've been running a lot of sub tests off master (thus the fixes you've 
seen....), but I haven't had this one locally on master.

> I did some tests here and in the first version I forgot to add the rotational drag, that was corrected some time ago.
> https://github.com/ArduPilot/ardupilot/pull/10701/files#diff-6c79088e2de21c5e3ab9572d0f7b4096R83
> 
> With my tests with sitl and manual control via QGC I was able to see the drag with angular and linear velocities while testing this PR and the same thing
> is happening in my local tests.
> 
> Are this outputs and plots from travis or from your local test ?

I downloaded the logs from travis's upload to firwmare.ardupilot.org

> It's a bit late here and I'll not be able to test and take a look right now.
> I also need to finish a paper until sunday, but I can take a look if you point how can I replicate this behaviors.

I'll see if I can work out how to replicate...

**Comment by peterbarker** [comment]
@patrickelectric It can be reproduced on trusty64; try using the Vagrant VM to do so.

master passes.  This PR doesn't.

Intriguing.

File: libraries/SITL/SIM_Submarine.h [hunk 2]
```diff
@@ -50,9 +50,24 @@ class Submarine : public Aircraft {
...
+        // Calculate total volume from water buoyancy
+        // $ V = F_b / (rho * g) $
+        // V = volume (m^3), rho = water density (kg/m^3), g = gravity (m/s^2), F_b = force (N)
+        float volume = buoyancy_acceleration * weight / (GRAVITY_MSS * 1023.6f);
+        // Calculate equivalent sphere area for drag force
+        // $ A = pi * r^2 / 4 $
+        // $ V = 4 * pi * r^3 / 3 $
+        // $ r ^2 = (V * 3 / 4) ^ (2/3) $
+        // A = area (m^3), r = sphere radius (m)
+        float equivalent_sphere_area = M_PI_4 * pow(volume * 3 / 4, 2 / 3);
```
Comment (peterbarker) [inline libraries/SITL/SIM_Submarine.h]: This is going to be *a* problem; you're doing integer arithmetic here, resulting in a lot of zeroes (and `pow(...)` returning 1...

Note sure if it is going to be *the* problem...

**Comment by patrickelectric** [comment]
@peterbarker 	Fixed, it was the numeric problem that you point related to int division and the angular drag factor (rad/s).

**Comment by peterbarker** [comment]
Merged, thanks!

... and now the real worry: WHY DID THIS WORK ON ANYTHING?!

How many lurking bugs do we have like this?

**Comment by patrickelectric** [comment]
Good question.. it was only possible to replicate it in trusty:14.04 (tested with docker locally)

**Comment by peterbarker** [comment]
@patrickelectric Well, the point here is that we were able to replicate the *bug* with everything *except* trusty64, right?  There should be no implicit conversion of the integers into floats - that's just wrong as you may really have meant to do that....

**Pull request merged**